File: python/ch03_agent/agent_ch03/bootstrap.py
"""组合根：把外部实现装配为第 3 章 Agent。

这是什么：依赖注入容器，负责创建和连接各层对象。
Java 类比：类似 Spring 的 @Configuration 类或 ApplicationContext。
为什么需要：单一职责原则，装配逻辑和业务逻辑分离，便于测试时替换依赖。

Java 对照：可以把本文件理解为 Spring 的 `@Configuration` 类。
它负责创建对象和连接依赖，但不负责真正的 Agent 业务流程。
"""

from .adapters.filesystem import LocalWorkspaceFileSystem
from .adapters.powershell import PowerShellRunner
from .core.commands import CommandRunner
from .core.filesystem import WorkspaceFileSystem
from .core.loop import AgentRunner, ToolAuthorizer
from .core.model import ModelClient
from .core.permissions import (
    ApprovalProvider,
    AuditSink,
    PermissionPolicy,
    PermissionRule,
)
from .core.profiles import P01, P02, P03, ChapterProfile
from .features.builtin_tools import create_chapter_one_tools, create_chapter_two_tools

SYSTEM_PROMPT = "You are a coding agent. Use tools when needed, inspect their results, and answer accurately."
# Java 对照：上面是模块常量，作用近似 `private static final String SYSTEM_PROMPT`。


def build_agent(profile: ChapterProfile, model: ModelClient, workspace: str, command_runner: CommandRunner | None = None, file_system: WorkspaceFileSystem | None = None, authorizer: ToolAuthorizer | None = None, approval_provider: ApprovalProvider | None = None, audit_sink: AuditSink | None = None, max_turns: int = 20) -> AgentRunner:
    """创建一个第 3 章 AgentRunner。

    `command_runner or PowerShellRunner()` 表示：调用者传了 Fake 就用 Fake，
    没传时才创建真实 PowerShellRunner。这就是最简单的依赖注入。
    """
    # `is` 比较对象身份，不是字段值。这样调用方不能 new 一个内容相同的 DTO
    # 冒充固定章节配置，Java 中相当于只接受预定义的单例常量。
    if profile is not P01 and profile is not P02 and profile is not P03:
        raise ValueError("必须传入固定的章节配置对象")
    command = command_runner or PowerShellRunner()
    tools = (
        create_chapter_one_tools(command)
        if profile.chapter == 1
        else create_chapter_two_tools(command, file_system or LocalWorkspaceFileSystem())
    )
    policy = None
    if profile is P03:
        if approval_provider is None:
            raise ValueError("第三章必须提供 approval_provider")
        if audit_sink is None:
            raise ValueError("第三章必须提供 audit_sink")
        policy = PermissionPolicy(
            rules=(
                PermissionRule(
                    "confirm-file-write",
                    "ask",
                    "第三章的文件写入需要明确审批",
                    lambda request: request.prepared.definition is not None
                    and request.prepared.definition.name in {"write_file", "edit_file"},
                ),
            ),
            approval=approval_provider,
            audit=audit_sink,
            write_boundary=file_system or LocalWorkspaceFileSystem(),
        )
    return AgentRunner(model, tools, SYSTEM_PROMPT, workspace, max_turns=max_turns, authorizer=authorizer, permission_policy=policy)

File: typescript/chapters/ch03/src/bootstrap.ts
// 组合根：按章节能力选择基础设施、工具集和权限策略。
import { NodeWorkspaceFileSystem } from "./adapters/filesystem.js";
import { PowerShellRunner } from "./adapters/powershell.js";
import type { CommandRunner } from "./core/commands.js";
import type { WorkspaceFileSystem } from "./core/filesystem.js";
import { AgentRunner } from "./core/loop.js";
import type { ModelClient } from "./core/model.js";
import type { ApprovalProvider, AuditSink } from "./core/permissions.js";
import { PermissionPolicy, PermissionRule } from "./core/permissions.js";
import type { ChapterProfile } from "./core/profiles.js";
import { profileForChapter } from "./core/profiles.js";
import { createChapterOneTools, createChapterTwoTools } from "./features/builtin-tools.js";

// 第 3 章把权限策略接入既有循环，写操作和 Shell 执行均经过审批。
const SYSTEM_PROMPT =
  "You are a coding agent. Use tools when needed, inspect their results, and answer accurately.";

export interface BuildDependencies {
  // 外部边界均可注入，离线测试无需启动真实进程或网络客户端。
  readonly model: ModelClient;
  // 命令和文件工具共享的工作区根。
  readonly workspace: string;
  // 可替换命令执行边界。
  readonly commandRunner?: CommandRunner;
  // 可替换文件系统边界，同时用于写路径安全校验。
  readonly fileSystem?: WorkspaceFileSystem;
  // P03 及以后必需的 ask 决策收敛边界。
  readonly approvalProvider?: ApprovalProvider;
  // P03 及以后必需的最终决策审计边界。
  readonly auditSink?: AuditSink;
  // 可选模型回合上限。
  readonly maxTurns?: number;
}

// 根据固定章节 profile 组合累计工具与权限策略，拒绝同号伪造 profile。
export function buildAgent(profile: ChapterProfile, dependencies: BuildDependencies): AgentRunner {
  // 禁止调用方伪造同章节号但能力不同的 profile，保持教学快照固定。
  if (profileForChapter(profile.chapter) !== profile) {
    throw new Error("profile must be a fixed chapter profile");
  }
  const commandRunner =
    dependencies.commandRunner === undefined ? new PowerShellRunner() : dependencies.commandRunner;
  const fileSystem =
    dependencies.fileSystem === undefined ? new NodeWorkspaceFileSystem() : dependencies.fileSystem;
  const tools =
    profile.chapter === 1
      ? createChapterOneTools(commandRunner)
      : createChapterTwoTools(commandRunner, fileSystem);
  const permissionPolicy = permissionPolicyForProfile(profile, fileSystem, dependencies);
  return new AgentRunner({
    model: dependencies.model,
    tools,
    systemPrompt: SYSTEM_PROMPT,
    workspace: dependencies.workspace,
    ...(permissionPolicy === undefined ? {} : { permissionPolicy }),
    ...(dependencies.maxTurns === undefined ? {} : { maxTurns: dependencies.maxTurns }),
  });
}

// 为具 policy 能力的章节构造完整策略；旧 profile 保持与前章一致的直接执行路径。
function permissionPolicyForProfile(
  profile: ChapterProfile,
  fileSystem: WorkspaceFileSystem,
  dependencies: BuildDependencies,
): PermissionPolicy | undefined {
  // P03 起审批与审计是必需依赖；更早 profile 保持原有无策略运行方式。
  if (!profile.capabilities.has("policy")) {
    return dependencies.approvalProvider === undefined
      ? undefined
      : new PermissionPolicy({ approval: dependencies.approvalProvider });
  }
  if (dependencies.approvalProvider === undefined) {
    throw new Error("approvalProvider is required for chapter 3 or later");
  }
  if (dependencies.auditSink === undefined) {
    throw new Error("auditSink is required for chapter 3 or later");
  }
  return new PermissionPolicy({
    rules: [
      new PermissionRule({
        name: "confirm-file-write",
        behavior: "ask",
        reason: "File writes require explicit approval from chapter 3 onward",
        matches: (request) => {
          // 只提升写入工具；Shell 的执行审批由策略默认规则统一处理。
          const name = request.prepared.definition?.name;
          return name === "write_file" || name === "edit_file";
        },
      }),
    ],
    approval: dependencies.approvalProvider,
    audit: dependencies.auditSink,
    writeBoundary: fileSystem,
  });
}
